### ISSUE TITLE
compacting loom on boot -> bail meme repeatedly

### ISSUE THREAD

**Issue by matildepark** [urbit/urbit#2677]
What good timing!

```
matilde@matilde:~$ ./urbit haddef-sigwen/
~
urbit 0.10.4
boot: home is /home/matilde/haddef-sigwen
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
boot: protected loom
live: loaded: GB/1.602.945.024
boot: installed 262 jets

work: contiguous free space: MB/544.569.464
work: measuring memory:
work: play 5352376
---------------- playback starting----------------
pier: replaying events 5352376 through 5352389
  kernel: GB/1.224.866.996
total arvo stuff: GB/1.224.866.996
  warm jet state: KB/96.900
  cold jet state: MB/149.833.456
  hank cache: B/372
  battery hash cache: B/288
  call site cache: B/100
  hot jet state: KB/127.744
total jet stuff: MB/150.058.860
  bytecode programs: MB/1.727.192
  bytecode cache: KB/212.432
total nock stuff: MB/1.939.624
  trace buffer: B/36
  memoization cache: B/288
total road stuff: B/324
total marked: GB/1.376.865.804
free lists: MB/226.046.960
sweep: GB/1.376.865.804

.............work: measuring memory:
.
---------------- playback complete----------------
ames: live on 0
  kernel: GB/1.231.381.740
total arvo stuff: GB/1.231.381.740
  warm jet state: KB/96.900
  cold jet state: MB/149.833.456
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: KB/127.744
total jet stuff: MB/150.058.676
  bytecode cache: B/288
total nock stuff: B/288
  trace buffer: B/36
  memoization cache: B/288
total road stuff: B/324
total marked: GB/1.381.441.028
free lists: MB/228.918.312
sweep: GB/1.381.441.028

work: compacting loom
bail: meme
bailing out
pier: work error: end of file
pier: work exit: status 0, signal 6
```

**Comment by joemfb** [comment]
I wish I knew what was using so much memory in your kernel instance. Unfortunately, we don't yet have tools to get past this state. ("compacting loom" is `|pack`, which is the best we have. We need a version that's read-only wrt the current loom.)

I'm afraid a personal breach is your only option at this time.

**Comment by matildepark** [comment]
I’m replaying checkpoint but … if it keeps doing it, I’ll breach. RIP me

—
~haddef-sigwen
https://urbit.org

**Comment by joemfb** [comment]
If you can force a `|pack` earlier in your ship's history, that might get you unstuck. That'll require a custom binary -- I can help you hack something together.

**Comment by matildepark** [comment]
It hung around a fifth of the way through the replay; it continues if I restart it, but I figure if I'm going to do this, I should probably have that early `|pack`. What should I be doing? Presumably making some changes somewhere, compiling with `make` and tar + scp-ing it over ...

**Comment by matildepark** [comment]
Time was of the essence, so I just breached; it would probably good to have these instructions somewhere anyhow...

**Comment by rabsef-bicrym** [comment]
I would like to try the described path if someone could give me instructions - rabsef breaching (again) would make me sad - I'd been doing so well lately

**Comment by BrienHartung** [comment]
This same thing has happened to me:

[urbit-planet-dump.txt](https://github.com/urbit/urbit/files/4488448/urbit-planet-dump.txt)

`
root@divmec-panfex-ubuntu:~/urbit-folder# free -h
              total        used        free      shared  buff/cache   available
Mem:          1.9Gi       122Mi       1.7Gi       0.0Ki        90Mi       1.7Gi
Swap:         2.0Gi        64Mi       1.9Gi
root@divmec-panfex-ubuntu:~/urbit-folder# df -H /home
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        52G   15G   38G  29% /
root@divmec-panfex-ubuntu:~/urbit-folder# 
`

**Comment by joemfb** [comment]
Sorry for the delay!

I'm working on some changes that should make in this much easier to control. In the meantime, working around it will require hacking on the worker process.

Here's the code that run memory compaction after successful events: https://github.com/urbit/urbit/blob/18878727b0f75c1c0f80cf518fa593b3ada788f5/pkg/urbit/worker/main.c#L600-L602. The gating condition (`pac_o`) is set in a couple of situations: memory pressure thresholds, or the presence of the `[%pack ~]` effect (produced by `|pack`). Those could be changed or disabled, or more conditions could be added.

The failure here is that `|pack` doesn't work past a certain level of memory usage. The "fix" I'm proposing is just to run it before memory usage hits those levels (unfortunately, I'm not sure what those levels are, there are several factors). The easiest way to do this would just be to replay the whole log and run `|pack` at a certain event number, or every N events. We already do a similar thing for cache reclamation: https://github.com/urbit/urbit/blob/18878727b0f75c1c0f80cf518fa593b3ada788f5/pkg/urbit/worker/main.c#L570-L578.

Unfortunately, replay and `|pack` are both slow as implemented, so this will take awhile. LMK if you have questions.

**Comment by joemfb** [comment]
To replay the full log, you'll have to delete the memory-image snapshot (the pier `$pier/.urb/chk/` directory). When you restart urbit, it will automatically start playing the log from the beginning.

**Comment by kennyrowe** [comment]
I tried this hack but looks like `+zuse` is not available?

```
urbit 0.10.4
boot: home is /root/new_urbit/sicdev-pilnup
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
work: play 1
---------------- playback starting----------------
pier: replaying events 1 through 9394389
work: (1)| boot
work: (2)| boot
work: (3)| boot
work: (4)| boot
work: (5)| boot
work: (5)| pill: 2a4f7637
fatal error: +zuse not present
pier: work error: end of file
pier: work exit: status 1, signal 0
```

**Comment by joemfb** [comment]
@kennyrowe, I think you're using an unreleased binary. That error message comes from some version negotiation that was temporarily added, then reverted. You should be fine to switch to the released v0.10.4 binaries.

/cc @jtobin

**Comment by kennyrowe** [comment]
@joemfb yeah you're right it was an RC binary. I ended up downloading the pier to a beefier machine anyway in hopes of processing the events faster.

9394389 need to be replayed for my ship so I'm guessing it's going to take about 17 or 18 hours in total. 

I will check in tomorrow with my results

**Comment by gusmacaulay** [comment]
I tried this with the 0.10.4 binary, deleting the chk directory.  It failed after a few events;
urbit 0.10.4
boot: home is /home/gus/Source/URBIT/lomped-firser
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
work: play 1
---------------- playback starting----------------
pier: replaying events 1 through 4329133
work: (1)| boot
work: (2)| boot
work: (3)| boot
work: (4)| boot
work: (5)| boot
work: (5)| pill: 480dbadc
work: (5)| core: 2572f150
.....work: fail: bad jar
work: invalid %work for event 10.
work: computed mug is 834ade8 but event 10 expected 5c708943.
pier: work error: end of file
pier: work exit: status 1, signal 0

**Comment by joemfb** [comment]
Hey, @gusmacaulay, I'm sorry for the delay.

That error is quite surprising, I've only seen a couple mug (short-hash) mismatches and never so early in the boot process. It'd be very helpful if you could make a tarball of your pier and send it to support @ urbit.org so we can investigate further.

In the meantime, to continue with replay, you'll need to modify the binary and disable that mug check here:

https://github.com/urbit/urbit/blob/ea9fdbbef37a1426acb51159bab87200c6b5dfe8/pkg/urbit/worker/main.c#L768-L776

Given that you were already facing the bail:meme issue, it'd probably be easier to perform an individual breach. But if there's data on that ship you want to retrieve, you can disable the mug checks, disable `|pack` (or run it more frequently, say every 10K events) as previously discussed, and then replay the log.

Improvements are coming this area, but they're still a ways out. I'm sorry for the difficulties!

**Comment by kennyrowe** [comment]
@joemfb what modifications would you suggest to `main.c`? 

I can't read c to save my life, but I can copy paste and build from source.

**Comment by joemfb** [comment]
I've pushed two branches: `v0.10-recovery` and `v0.10-replay-recovery`. Once their CI builds finish, I'll post links to the static binaries that will be built.

Use the first to just try to keep running, and the second if you intend to replay your whole event-log. Both branches disable enforcement of the mug-check, the first completely disables auto-pack (but explicit `|pack` will still work), the second runs auto-pack every 20k events.

Again, sorry for all the difficulties in this area.

**Comment by joemfb** [comment]
Links:

`v0.10-recovery`:

- linux: //bootstrap.urbit.org/99f294c06153f1a5210f59f7382d38451410cdee-linux64.tgz 
- macos: //bootstrap.urbit.org/99f294c06153f1a5210f59f7382d38451410cdee-darwin.tgz 

`v0.10-replay-recovery`:

- linux: //bootstrap.urbit.org/ef938b7a18a60abcde4d1872d07e5adc926ee570-linux64.tgz 
- macos: //bootstrap.urbit.org/ef938b7a18a60abcde4d1872d07e5adc926ee570-darwin.tgz 

/cc @jtobin, the naming convention change didn't account for the per-commit binaries, it'd be nice to get "urbit" back into these filenames.

**Comment by kennyrowe** [comment]
@joemfb worked!

**Comment by joemfb** [comment]
Glad you're back! Can you run `|mass` and save the output to a pastebin? It'd be good to know where the memory usage is.

**Comment by kennyrowe** [comment]
Here you go

https://pastebin.com/LSnbdNGC

**Comment by joemfb** [comment]
Wow, that's very modest usage! I have no explanation for why it was failing.

**Comment by gusmacaulay** [comment]
I copied back my original chk directory and used the recovery binary - it worked!  Starting to wondering if my hardware is cursed. I have a copy of the broken pier if you still want to look at it.  |mass is here https://pastebin.com/aryuxfdz

**Comment by t-e-r-m** [comment]
I replayed the entire log using `v0.10-replay-recovery`, and it reduced memory usage by about half a gigabyte for me. Planet with ~2.3M events. Thanks @joemfb

**Comment by gusmacaulay** [comment]
After a few days of operation the planet crashed again with a new error I hadn't noticed previously "old and new builds have same date, will probably crash!" ... and it did! See here for log from just before crash https://pastebin.com/1ge85X9q 

I tried restarting with v0.10-recovery as before but to no avail - playback completes but then i get dreaded BMOD (bail meme of death)  
|«behn»
bail: meme
bailing out.

**Comment by matildepark** [comment]
Have been trying to help ~fallyn-balfus with this issue for, I guess a week now ... was replaying the event log from scratch with replay-recovery. The instance seems to provide mug mismatch now:

```
~
urbit 0.10.4
boot: home is /home/urb/fallyn-balfus
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/217.202.688
boot: installed 262 jets
work: play 1985002
---------------- playback starting----------------
pier: replaying events 1985002 through 3483517
work: computed mug is 5e949643 but event 1985002 expected 4480ccea.
work: computed mug is 634c205e but event 1985003 expected 533654c.
work: computed mug is 6de10b13 but event 1985004 expected 6a3548b3.
work: computed mug is 22e9b206 but event 1985005 expected 7d1f3c41.
work: computed mug is 49dec0b0 but event 1985006 expected 7f346569.
work: computed mug is 54665118 but event 1985007 expected 532a2e13.
work: computed mug is 5c49a4a8 but event 1985008 expected 46f9995c.
work: computed mug is 1ad4d44c but event 1985009 expected 16d4a490.
work: computed mug is 4bfe640 but event 1985010 expected 1ddede02.
work: computed mug is 2b085b47 but event 1985011 expected 57b8c8e3.
work: computed mug is 7979d7e9 but event 1985012 expected 72103145.
work: computed mug is 1c84c260 but event 1985013 expected 56958a7b.
```

The replay is also very, very slow. Unsure. I deleted chk and started replaying it again. I am wondering if we are left with a breach at this point.

**Comment by gusmacaulay** [comment]
@matildepark I am also attempting replay-recovery option after my last crash (see above) - same issue as you report, mug events not matching at the start, and replay takes a long time (started yesterday and not complete yet).

**Comment by joemfb** [comment]
The mug mismatches are due to #2396 -- very frustrating, but not a fatal error (on that branch). Replay is already horribly slow in the current version, and `replay-recovery` is going to even slower due to its frequently running `|pack`. These issues are my top priority!

But in the meantime, I can only apologize for these difficulties and ask for your patience. You can always breach if you're unwilling to wait for the replay. If you do, you should manually `|pack` after your ship has applied the OTA updates.

**Comment by gusmacaulay** [comment]
Thanks @joemfb you'll be pleased to know my ship did eventually complete the replay and is alive again.

**Comment by vvisigoth** [comment]
NICE!
—
~poldec-tonteg
http://urbit.org

**Comment by joemfb** [comment]
@gusmacaulay, that's great to hear -- welcome back!

**Comment by kennyrowe** [comment]
~sicdev-pilnup died again. Trying to replay again but not sure I can wait for 13m events. Might need to breach.

**Comment by joemfb** [comment]
Sorry to hear that, @kennyrowe. If you plan to breach, we should coordinate first.

@philipcmonk, this is probably relevant to the %ames fixes in #2852, not sure how we plan to go about getting those out.

**Comment by kennyrowe** [comment]
Replaying now.  hoping to get through the log in a few days

**Comment by gusmacaulay** [comment]
@kennyrowe for reference my planet replayed 4.8m events over (I think) ~30 hours or so - a few days sounds about right for 13m events

**Comment by gusmacaulay** [comment]
It's happening again.  This time it gracefully/kindly provided bak files for north.bin and south.bin.  I copied them back and restarted but it fails after playing the remaining events and compacting loom - https://pastebin.com/RciWzVgB - still using the replay-recovery binary.

**Comment by joemfb** [comment]
@gusmacaulay, you should switch to the "recovery" branch with `|pack` disabled. There might be other things you can do to reclaim some of that space, but those are blocked on being-running, and if `|pack` is failing the only way forward is to disable it (at least temporarily).

**Comment by gusmacaulay** [comment]
@joemfb I switched to the 'recovery' build again and the same thing happened. Is |pack disabled in that build, or are you saying I need to disable it somehow?

**Comment by kennyrowe** [comment]
After replaying my entire event log again (13m) I got the same `pier: work exit: status 0, signal 6` error. I was about to breach but on a whim, I checked the event stream folder.

I noticed that in `sicdev-pilnup/.urb/chk` there were two new files `north.bin.bak and south.bin.bak` I guessed these might be backups of the event log so I copied them to `north.bin and south.bin`

After that, I booted with the recovery binary (linked above) and it worked. Not sure if that was intentional or not @joemfb but it was nice feature.

**Comment by gusmacaulay** [comment]
After switching back to replay-recovery and replaying my event log again I got the same signal 6 error.  I tried copying the bak files back over but after the catchup replay it still failed.  I also tried shifting my pier to a different disk, but still no good........all these events ... like tears in the rain .... time to breach

**Comment by benjaminkwilliams** [comment]
This hit me this morning. 
![pack_bail_meme](https://user-images.githubusercontent.com/1291662/83142700-796e8680-a0b6-11ea-9450-0e2a7176e2b5.png)
are the binaries above still what I need to run?

**Comment by somlus-savlev** [comment]
Hey there, somlus-savlev here. Had this issue, and unfortunately a replay will require about 200,000,000 events. I'm assuming that no matter what I would have to breach given the extreme time investment for that. would you agree?  I would love to somehow scrape the Art By Urbit group postings: Sigils and UrbArt.  LMK if you think there is any way for me to get online enough to get these.  Thank goodness this is not on the hooniverse planet. I will stop doing |pack commands when things get irregular and sluggish.  Along those lines it would be nice to get a simple bag of harmless but maybe helpful tricks when performance is slow or chat interface is choppy.  I'm certainly aware that this request might be way oversimplifying things.

**Comment by philipcmonk** [comment]
If you have the .bak files, you should be able to follow the process here: https://github.com/urbit/urbit/issues/2677#issuecomment-627481880 using the binary you can download here (recovery):
https://github.com/urbit/urbit/issues/2677#issuecomment-617431649

**Comment by somlus-savlev** [comment]
Thank you, recovery complete.


Here is my |mass output

https://pastebin.com/EFi2kKVA

**Comment by forfel-norfel** [comment]
@joemfb Thanks for making those recovery builds available. They helped me recover my planet after I ran out of disk space.

I'm trying to recover from an unrelated issue now and I noticed that the replay version seems to keep all of the deleted north/south.bin.bak files open, so it slowly fills up the disk until it fails from a lack of disk space. Once the process terminates, there's lots of free disk space since the deleted files are able to be removed, but the most recent versions that are left might be garbage as a result of running out of space. It's not a big deal to babysit it by manually restarting it once or twice in this case because I don't have much to recover, but it seemed worth reporting. I'm using it on Linux if that matters.
